### DISCUSSION TITLE
Custom local namespace when running in Web Worker

### DISCUSSION THREAD

**Discussion by dmondev** [pyodide/pyodide#1382] [Q&A]
Hi, 

I'm trying to understand/use custom namespaces when running Pyodide inside a web worker.

My goal is to share some state between code executions, as well as having some fresh, not shared local variables for each execution. I'm trying not to use js or self for this.

When I' trying to use both global and local custom namespaces with eval_code, whatever I import outside of a function is no longer available See the first try catch block in the snippet below.

However, if I just remove the locals (second try/catch), it works. Is there a way to use custom namespace for locals, and still use the import syntax, local variables outside of the function ?

```javascript

let globals = pyodide.globals.get('dict')()
let locals = pyodide.globals.get('dict')()
let python = `
from js import console
def my_broken_function():
  console.log("I dont do much")

my_broken_function()

`
// This doesnt work (as I expect)
try {
  pyodide.pyodide_py.eval_code(python, globals, locals)
} catch(e) {
  console.error(`failed to run`, e);
}

// This works
try {
  pyodide.pyodide_py.eval_code(python, globals)
  console.log("I'm working !! Console is available")
} catch(e) {
  console.error(`failed to run`, e);
}

```

Thanks !!

**Comment by hoodmane** [comment]
Python scopes are really weird. If you read the fine print about `exec` it says:
>If exec gets two separate objects as globals and locals, the code will be executed as if it were embedded in a class definition.

What you expect is for it to be executed as if it's in it's own _function_, which would make way more sense.
This works:
```python
def g():
   x = 2
   def f():
      print(x)
   f()
g()
```
This doesn't work:

```py
class Test:
   x = 2
   def f():
      print(x)
   f()
```
Aggravatingly,
```py
exec(
"""
x = 2
def f():
   print(x)
f()
""",
{}, {})
```
is equivalent to the second code block. I think there is basically no way to access the variable `x` in this case. I am not really sure why `exec` works this way, it makes the `locals` argument seem pretty not that useful.

**Reply by dmondev** [reply]
Thanks for the reply, that clarifies it !!
